ui/src/app/job-details/resources/resources.component.ts

```diff
@@ -0,0 +1,138 @@
+import {
+  Component,
+  Input,
+  OnInit,
+  ViewContainerRef,
+} from '@angular/core';
+import {
+  MatExpansionPanel,
+  MatSnackBar,
+  MatSnackBarConfig,
+  MatTabChangeEvent
+} from '@angular/material'
+
+import {ErrorMessageFormatterPipe} from '../../shared/pipes/error-message-formatter.pipe';
+import {JobMetadataResponse} from '../../shared/model/JobMetadataResponse';
+import {ResourceUtils} from '../../shared/utils/resource-utils';
+import {GcsService} from '../../core/gcs.service';
+
+@Component({
+  selector: 'jm-resources',
+  templateUrl: './resources.component.html',
+  styleUrls: ['./resources.component.css'],
+})
+export class JobResourcesComponent implements OnInit {
+  @Input() job: JobMetadataResponse;
+
+  sourceFile: string = '';
+  inputs: Array<string> = [];
+  outputs: Array<string> = [];
+  logFileData: Map<string, string> = new Map();
+
+  tabIds: Array<string> = [];
+  tabTitles: Map<string, string> = new Map([
+    ['inputs', 'Inputs'],
+    ['outputs', 'Outputs'],
+    ['source-file', 'Source File']
+  ]);
+  currentTabId: string;
+
+  constructor(
+    private readonly gcsService: GcsService,
+    private errorBar: MatSnackBar,
+    private readonly viewContainer: ViewContainerRef) {}
+
+  ngOnInit() {
+    if (this.job.inputs) {
+      this.inputs = Object.keys(this.job.inputs).sort();
+      this.tabIds.push('inputs');
+    }
+    if (this.job.outputs) {
+      this.outputs = Object.keys(this.job.outputs).sort();
+      this.tabIds.push('outputs');
+    }
+
+    if (this.job.extensions) {
+      if (this.job.extensions.sourceFile) {
+        this.sourceFile = this.job.extensions.sourceFile;
+        this.tabIds.push('source-file');
+      }
+
+      if (this.job.extensions.logs) {
+        for (let file of Object.keys(this.job.extensions.logs)) {
+          this.readResourceFile(file);
+          let tabId = 'log-' + file;
+          this.tabIds.push(tabId);
+          this.tabTitles.set(tabId, file);
+        }
+      }
+    }
+
+    if (this.tabIds.length > 0) {
+      this.currentTabId = this.tabIds[0];
+    }
+  }
+
+  alwaysExpanded(): boolean {
+    return !this.job.extensions || !this.job.extensions.tasks;
+  }
+
+  tabs(): string[] {
+    return this.tabIds.map(id => this.tabTitles.get(id));
+  }
+
+  logFiles(): string[] {
+    return Array.from(this.logFileData.keys());
+  }
+
+  tabChanged(event: MatTabChangeEvent) {
+    this.currentTabId = this.tabIdForTitle(event.tab.textLabel);
```

Comment:
I still don't think this problem is resolved, but let me know if I'm missing something. I had meant to imply that you would start using the tab index here instead. The properties are a bit confusing, but I think `this.tabIds[event.tab.origin + event.tab.position]` gives you the right result?

Reply:
Good point, you're right that doesn't fix it since we're still looking up by title.

I'm not sure what `tab.origin` and `tab.position` are, but they do not seem to represent the index of the tab––curiously it almost always seems to be 1 and 0, respectively and sometimes -1 and 0.

Luckily this is actually available on the event, `event.index`, so switched to that.

Reply:
Oops - I was looking at `MatTab` rather than `MatTabChangeEvent`. Nice, thanks!